pkg/auth/db_service.go

```diff
@@ -867,3 +882,304 @@ func (s *DBAuthService) markTokenSingleUse(ctx context.Context, tokenID string,
+		}
+		model.ConvertGroup(&dbGroup)
+		key := model.KVGroupPath(dbGroup.DisplayName)
+		value, err := proto.Marshal(model.ProtoFromGroup(kvGroup))
+		if err != nil {
+			return nil, err
+		}
+		if err = je.Encode(kv.Entry{
+			PartitionKey: []byte(model.PartitionKey),
+			Key:          []byte(key),
+			Value:        value,
+		}); err != nil {
+			return nil, err
+		}
+	}
+	return groupID2Name, nil
+}
+
+func exportPolicies(ctx context.Context, d *pgxpool.Pool, je *json.Encoder) (IDToName, error) {
+	rows, err := d.Query(ctx, "SELECT * FROM auth_policies ORDER BY created_at ASC")
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	policyID2Name := make(IDToName)
+	scanner := pgxscan.NewRowScanner(rows)
+	for rows.Next() {
+		dbPolicy := model.DBPolicy{}
+		err = scanner.Scan(&dbPolicy)
+		if err != nil {
+			return nil, err
+		}
+		key := model.KVPolicyPath(dbPolicy.DisplayName)
+		value, err := proto.Marshal(model.ProtoFromPolicy(&dbPolicy.BasePolicy, strconv.Itoa(dbPolicy.ID)))
+		if err != nil {
+			return nil, err
+		}
+		if err = je.Encode(kv.Entry{
+			PartitionKey: []byte(model.PartitionKey),
+			Key:          []byte(key),
+			Value:        value,
+		}); err != nil {
+			return nil, err
+		}
+		policyID2Name[dbPolicy.ID] = dbPolicy.DisplayName
+	}
+	return policyID2Name, nil
+}
+
+func exportUserGroups(ctx context.Context, d *pgxpool.Pool, je *json.Encoder, usersDetails UserIDToDetails, groupsNames IDToName) error {
+	type relation struct {
+		UserID  int64 `db:"user_id"`
+		GroupID int   `db:"group_id"`
+	}
+	rows, err := d.Query(ctx, "SELECT * FROM auth_user_groups")
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+	scanner := pgxscan.NewRowScanner(rows)
+	for rows.Next() {
+		ug := relation{}
+		err = scanner.Scan(&ug)
+		if err != nil {
+			return err
+		}
+		username := usersDetails[ug.UserID].name
+		key := model.KVUserToGroup(username, groupsNames[ug.GroupID])
+		secIndex := kv.SecondaryIndex{PrimaryKey: []byte(model.KVUserPath(username))}
+		value, err := proto.Marshal(&secIndex)
+		if err != nil {
+			return err
+		}
+		if err = je.Encode(kv.Entry{
+			PartitionKey: []byte(model.PartitionKey),
+			Key:          []byte(key),
+			Value:        value,
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func exportUserPolicies(ctx context.Context, d *pgxpool.Pool, je *json.Encoder, usersDetails UserIDToDetails, policiesNames IDToName) error {
+	type relation struct {
+		UserID   int64 `db:"user_id"`
+		PolicyID int   `db:"policy_id"`
+	}
+	rows, err := d.Query(ctx, "SELECT * FROM auth_user_policies")
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+	scanner := pgxscan.NewRowScanner(rows)
+	for rows.Next() {
+		up := relation{}
+		err = scanner.Scan(&up)
+		if err != nil {
+			return err
+		}
+		username := usersDetails[up.UserID].name
+		key := model.KVPolicyToUser(policiesNames[up.PolicyID], username)
+		secIndex := kv.SecondaryIndex{PrimaryKey: []byte(model.KVUserPath(username))}
```

Comment:
I think the key is of the form `users\<user>\policies\<policy>`, right?
So it should point to the policy and not the user then

Reply:
Right. Fixed